### PR TITLE
Call `Op.make_thunk` before generating the C code.

### PR DESCRIPTION
**Pull request by lamblin**
This should fix the compilation of `cuda_convnet`, and maybe also `cuda_ndarray`.
Fixes https://github.com/lisa-lab/pylearn2/issues/1463

### CODE REVIEW

File: theano/compile/debugmode.py [hunk 1]
```diff
@@ -1663,6 +1663,45 @@ def make_all(self, profiler=None, input_storage=None
...
+                #
+                # Note RP: this warp_thunk doesn't work. What happens is
+                # that for all ops that have a make_thunk, the same instance
+                # of `wrap_thunk` gets used ( that has the same `thunk`
+                # function, probably the one of the first of all those ops (
+                # or the last .. I'm not sure). I don't know suffcient about
+                # how python works to understand why. A bunch of tests fail
+                # because of this, one of them being
+                # theano/scan_module/tests/scan_tests.py:T_Scan.test_backwards
+                # def wrap_thunk():
```
Comment (abergeron) [inline theano/compile/debugmode.py]: If you want to make that work do it like this:
```
def wrap_thunk(thunk=thunk):
```

Comment (lamblin) [inline theano/compile/debugmode.py]: I just moved that part around, I have no idea what it was supposed to do or how it should have been used.

Comment (nouiz) [inline theano/compile/debugmode.py]: Personnaly, I would remove all that dead code and comments that you moved. The code is totally useless and the comments is about possible futur hypothesys we do not want. But as this PR make Theano better, I'll merge it now.

Comment (lamblin) [inline theano/compile/debugmode.py]: Addressed in #2742

**Pull request merged**